src/arkode/arkode_arkstep.c

```diff
@@ -3146,6 +3418,266 @@ int arkStep_ComputeSolutions_MassFixed(ARKodeMem ark_mem, sunrealtype* dsmPtr)
   return (ARK_SUCCESS);
 }
 
+/*---------------------------------------------------------------
+  Utility routines for interfacing with SUNAdjointStepper
+  ---------------------------------------------------------------*/
+
+int arkStep_fe_Adj(sunrealtype t, N_Vector sens_partial_stage,
+                   N_Vector sens_complete_stage, void* content)
+{
+  SUNErrCode errcode = SUN_SUCCESS;
+
+  SUNAdjointStepper adj_stepper           = (SUNAdjointStepper)content;
+  SUNAdjointCheckpointScheme check_scheme = adj_stepper->checkpoint_scheme;
+  ARKodeMem ark_mem = (ARKodeMem)adj_stepper->adj_sunstepper->content;
+  void* user_data   = adj_stepper->user_data;
+
+  N_Vector Lambda_part     = N_VGetSubvector_ManyVector(sens_partial_stage, 0);
+  N_Vector Lambda          = N_VGetSubvector_ManyVector(sens_complete_stage, 0);
+  N_Vector checkpoint      = N_VGetSubvector_ManyVector(ark_mem->tempv2, 0);
+  sunrealtype checkpoint_t = SUN_RCONST(0.0);
+
+  errcode = SUNAdjointCheckpointScheme_LoadVector(check_scheme,
+                                                  adj_stepper->step_idx,
+                                                  ark_mem->adj_stage_idx, 0,
+                                                  &checkpoint, &checkpoint_t);
+
+  // Checkpoint was not found, recompute the missing step
+  if (errcode == SUN_ERR_CHECKPOINT_NOT_FOUND) { return +1; }
+
+  if (adj_stepper->JacFn)
+  {
+    adj_stepper->JacFn(t, checkpoint, NULL, adj_stepper->Jac, user_data, NULL,
+                       NULL, NULL);
+    adj_stepper->njeval++;
+    if (SUNMatMatTransposeVec(adj_stepper->Jac, Lambda_part, Lambda))
+    {
+      return -1;
+    };
+  }
+  else if (adj_stepper->JvpFn)
+  {
+    adj_stepper->JvpFn(Lambda_part, Lambda, t, checkpoint, NULL, user_data, NULL);
+
+    adj_stepper->njtimesv++;
+  }
+  else if (adj_stepper->vJpFn)
+  {
+    adj_stepper->vJpFn(Lambda_part, Lambda, t, checkpoint, NULL, user_data, NULL);
+    adj_stepper->nvtimesj++;
+  }
+
+  if (adj_stepper->JacPFn)
+  {
+    if (N_VGetNumSubvectors_ManyVector(sens_complete_stage) < 2) { return -1; }
+    N_Vector nu = N_VGetSubvector_ManyVector(sens_complete_stage, 1);
+    adj_stepper->JacPFn(t, checkpoint, NULL, adj_stepper->JacP, user_data, NULL,
+                        NULL, NULL);
+    adj_stepper->njpeval++;
+    if (SUNMatMatTransposeVec(adj_stepper->JacP, Lambda_part, nu))
+    {
+      return -1;
+    }
+  }
+  else if (adj_stepper->JPvpFn)
+  {
+    if (N_VGetNumSubvectors_ManyVector(sens_complete_stage) < 2) { return -1; }
+    N_Vector nu = N_VGetSubvector_ManyVector(sens_complete_stage, 1);
+    adj_stepper->JPvpFn(Lambda_part, nu, t, checkpoint, NULL, user_data, NULL);
+    adj_stepper->njptimesv++;
+  }
+  else if (adj_stepper->vJPpFn)
```

Comment:
Did we settle on removing either `JPvpFn` or `vJPpFn`?

Reply:
I am removing vJPpFn.

Reply:
What do we think about still keeping the two different set functions:

```
SUNAdjointStepper_SetJacTimesVecFn
SUNAdjointStepper_SetVecTimesJacFn
```

but just having `SUNAdjointStepper_SetVecTimesJacFn` call `SUNAdjointStepper_SetJacTimesVecFn`. This would make it very clear to users that we support both options (it just doesnt matter to us implementation wise).